activerecord/lib/active_record/relation/calculations.rb

```diff
@@ -314,9 +312,7 @@ def execute_simple_calculation(operation, column_name, distinct) #:nodoc:
         result = skip_query_cache_if_necessary { @klass.connection.select_all(query_builder, nil) }
         row    = result.first
         value  = row && row.values.first
-        type   = result.column_types.fetch(column_alias) do
-          type_for(column_name)
-        end
+        type   = result.column_types.fetch(column_alias, Type.default_value)
```

Comment:
I find it strange that `column_alias` is nil here for count calculations with a subquery.

All in-tree adapters return integers from count queries, so how about dropping the type casting from that branch entirely?

```patch
diff --git a/activerecord/lib/active_record/relation/calculations.rb b/activerecord/lib/active_record/relation/calculations.rb
index f6d6821f72..658c9c84e8 100644
--- a/activerecord/lib/active_record/relation/calculations.rb
+++ b/activerecord/lib/active_record/relation/calculations.rb
@@ -291,6 +291,7 @@ def execute_simple_calculation(operation, column_name, distinct) #:nodoc:
           return 0 if limit_value == 0

           query_builder = build_count_subquery(spawn, column_name, distinct)
+          skip_query_cache_if_necessary { @klass.connection.select_value(query_builder) }
         else
           # PostgreSQL doesn't like ORDER BY when there are no GROUP BY
           relation = unscope(:order).distinct!(false)
@@ -307,14 +308,14 @@ def execute_simple_calculation(operation, column_name, distinct) #:nodoc:
           relation.select_values = [select_value]

           query_builder = relation.arel
+
+          result = skip_query_cache_if_necessary { @klass.connection.select_all(query_builder) }
+          row    = result.first
+          value  = row && row.values.first
+          type   = result.column_types.fetch(column_alias, Type.default_value)
+
+          type_cast_calculated_value(value, type, operation)
         end
-
-        result = skip_query_cache_if_necessary { @klass.connection.select_all(query_builder, nil) }
-        row    = result.first
-        value  = row && row.values.first
-        type   = result.column_types.fetch(column_alias, Type.default_value)
-
-        type_cast_calculated_value(value, type, operation)
       end

       def execute_grouped_calculation(operation, column_name, distinct) #:nodoc:
```

Reply:
👍 Updated.